File: java/ql/src/experimental/Security/CWE/CWE-348/ClientSuppliedIpUsedInSecurityCheck.java
import javax.servlet.http.HttpServletRequest;
import org.apache.commons.lang3.StringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Controller;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.ResponseBody;

@Controller
public class ClientSuppliedIpUsedInSecurityCheck {

    @Autowired
    private HttpServletRequest request;

    @GetMapping(value = "bad1")
    public void bad1(HttpServletRequest request) {
        String ip = getClientIP();
        if (!StringUtils.startsWith(ip, "192.168.")) {
            new Exception("ip illegal");
        }
    }

    @GetMapping(value = "bad2")
    public void bad2(HttpServletRequest request) {
        String ip = getClientIP();
        if (!"127.0.0.1".equals(ip)) {
            new Exception("ip illegal");
        }
    }

    @GetMapping(value = "good1")
    @ResponseBody
    public String good1(HttpServletRequest request) {
        String ip = request.getHeader("X-FORWARDED-FOR");
        // Good: if this application runs behind a reverse proxy it may append the real remote IP to the end of any client-supplied X-Forwarded-For header.
        ip = ip.split(",")[ip.split(",").length - 1];
        if (!StringUtils.startsWith(ip, "192.168.")) {
            new Exception("ip illegal");
        }
        return ip;
    }

    protected String getClientIP() {
        String xfHeader = request.getHeader("X-Forwarded-For");
        if (xfHeader == null) {
            return request.getRemoteAddr();
        }
        return xfHeader.split(",")[0];
    }
}

File: csharp/ql/test/experimental/Security Features/CWE-348/ClientSuppliedIpUsedInSecurityCheck.cs
using System;
using Microsoft.AspNetCore.Mvc;

namespace Testing.Controllers
{
    [Route("api/[controller]")]
    [ApiController]
    public class ClientSuppliedIpUsedInSecurityCheck : ControllerBase
    {
        [HttpGet("bad1")]
        public IActionResult Bad1()
        {
            var ip = GetClientIp();
            if (!ip.StartsWith("192.168."))
            {
                throw new Exception("illegal ip");
            }

            return Ok("Bad");
        }

        [HttpGet("bad2")]
        public IActionResult Bad2()
        {
            var ip = GetClientIp();
            if (!ip.Equals("127.0.0.1"))
            {
                throw new Exception("illegal ip");
            }

            return Ok("Bad");
        }

        [HttpGet("good1")]
        public IActionResult Good1()
        {
            string ip = Request.Headers["X-Forwarded-For"];
            // Good: if this application runs behind a reverse proxy it may append the real,
            // remote IP to the end of any client-supplied X-Forwarded-For header.
            ip = ip.Split(",")[ip.Split(",").Length - 1];

            if (!ip.StartsWith("192.168."))
            {
                throw new Exception("illegal ip");
            }

            return Ok();
        }

        private string GetClientIp()
        {
            var xfHeader = Request.Headers["X-Forwarded-For"];
            if (!string.IsNullOrEmpty(xfHeader))
            {
                return xfHeader.ToString().Split(",")[0];
            }

            xfHeader = Request.HttpContext
                .Connection.RemoteIpAddress?.ToString();

            if (string.IsNullOrEmpty(xfHeader))
            {
                throw new Exception("ip not found");
            }

            return xfHeader;
        }
    }
}
